WT-6616 Set the oldest timestamp of the checkpoint when it is finished

quchenhao opened the pull request.

agorrod commented:
@quchenhao please add a Python test case that demonstrates the bug.

quchenhao commented:
@agorrod Sure, I'll try that. However, I think it may be hard as we need to control eviction to run concurrently to checkpoint. I'm not sure that is currently possible in python test.

agorrod commented:
> I'm not sure that is currently possible in python test.

I'm happy to explore possibilities for making it possible if that will help.

quchenhao commented:
> > I'm not sure that is currently possible in python test.
> 
> I'm happy to explore possibilities for making it possible if that will help.

Thanks, I got an example from Hari. I think I'm on the track.

quchenhao commented:
@agorrod I managed to reproduce this in python but I'm concerned that the test may not be stable.

quchenhao commented on test/suite/test_checkpoint08.py:
```diff
@@ -0,0 +1,119 @@
+        try:
+            ckpt.start()
+            # Wait until checkpoint has get its timestamps
+            time.sleep(1)
```
I'm worried that this is not very reliable but I don't have a better solution.

sueloverso replied:
Your thought and comment is a good indicator that maybe a `test/csuite` program is the more natural place for a test instead of python. I suggest looking at `test/csuite/wt3363_checkpoint_op_races/main.c` for a test that might be a good starting place that should already have the infrastructure you need for this issue.

quchenhao replied:
Thanks Sue. I'll explore that.

quchenhao commented:
Unit tests all green in different build variants

sueloverso commented on test/suite/test_checkpoint08.py:
```diff
@@ -0,0 +1,118 @@
+        # Update the keys at timestamp 30
+        self.large_updates(uri, value2, nrows, 30)
+
+        # Remove the key at timestamp 40
```
Here

sueloverso commented on test/suite/test_checkpoint08.py:
```diff
@@ -0,0 +1,118 @@
+        # Set stable to timestamp 50.
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(50))
+
+        # Create a checkpoint thread
```
Here

sueloverso commented on test/suite/test_checkpoint08.py:
```diff
@@ -0,0 +1,118 @@
+        # Simulate a server crash and restart.
+        self.simulate_crash_restart(".", "RESTART")
+
+        # Check the data is present in the checkpoint after restart
```
Here.

quchenhao commented:
I replaced the python test with a csuite test. What it does is that it continuously insert keys and delete them. For each key, it is inserted at timestamp X and deleted at timestamp X+1. After each insertion, the stable timestamp is updated to X so checkpoint happens at X should contain everything from its oldest timestamp to X. We can verify that by reading the checkpoint after restart from oldest timestamp to X.

I think the change is ready for review now.

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,372 @@
+    "statistics=(fast),statistics_log=(wait=1,json=true),"    \
+    "timing_stress_for_test=[checkpoint_slow]"
+
+#define ENV_CONFIG_REC "log=(archive=false,recover=on)"
```
I see this is based of `timestamp_abort` which is good. However I see you removed the `log=(archive=true,file_max=10M,enabled),` settings from the configuration string. The `log_table` setting and the recovery configuration are not useful if logging is not enabled when the database is created.

quchenhao replied:
fixed.

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,372 @@
+    WT_CURSOR *cursor;
+    WT_ITEM data;
+    WT_SESSION *session;
+    uint64_t ts, oldest_ts;
```
Alphabetize.

quchenhao replied:
fixed

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,372 @@
+
+    /* Insert and then delete the keys until we're killed. */
+    printf("Worker thread started.\n");
+    for (ts = 1, oldest_ts = 0;; ++ts) {
```
Alphabetize here too.

Also, do you want to increment `ts` by 2 since you're using ts + 1 for the delete each loop? It seems like it is fine either way, I'm just wondering if there is any benefit one way or the other.

quchenhao replied:
I think both works the same.

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,372 @@
+
+    if (chdir(home) != 0)
+        testutil_die(errno, "parent chdir: %s", home);
+
```
I recommend re-instating the code to make a copy of the original database before recovery. It can be helpful looking at bugs (and that also means eventually you'll likely want the `-v` verify-only arg).

sueloverso commented:
@Chenhao-Huang can you tell that this test program is testing the scenario you want?

quchenhao commented:
> @Chenhao-Huang can you tell that this test program is testing the scenario you want?

Yes, I reproduced the issue with the test.

quchenhao commented on test/csuite/schema_abort/main.c:
```diff
@@ -1041,15 +1041,9 @@ main(int argc, char *argv[])
      */
     if (chdir(home) != 0)
         testutil_die(errno, "parent chdir: %s", home);
-    /*
```
The comment here is confusing. It is copying all the data not just the log. I noticed in random_abort we are only copying the log files. I changed it also to copy all the data as the other tests seems to be fine.

quchenhao commented:
The csuite test failed in one of the pr test:
https://spruce.mongodb.com/task/wiredtiger_ubuntu1804_csuite_wt6616_checkpoint_oldest_ts_test_patch_7a08dc67df07f3895410c4b6cf7a2c70aec3ce24_5f4858520ae60609a14839d2_20_08_28_01_05_23?execution=1

I think there is another problem here as the missing key is at the stable timestamp not at the oldest timestamp.

quchenhao commented:
This problem appears after I fixed the wiredtiger config @sueloverso pointed it is wrong. I can reproduce the problem locally. I think there is a problem with logged table.

kommiharibabu commented on src/include/txn.i:
```diff
@@ -510,15 +510,24 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
 
-    if (checkpoint_ts != 0 && checkpoint_ts < pinned_ts)
-        *pinned_tsp = checkpoint_ts;
+    if (checkpoint_ts != WT_TS_NONE && checkpoint_oldest_ts < pinned_ts) {
```
The check of NONE is against `checkpoint_ts` and comparison is with `checkpoint_oldest_ts`. Later the `checkpoint_ts` is used only to assert. These conditions can lead to confusion. Why can't we calculate the Minimum of both `checkpoint_ts` and `checkpoint_oldest_ts` and compare it against pinned_tst?

quchenhao replied:
We only need to pin the timestamp if the checkpoint uses timestamp and that is determined by checkpoint_ts. It's possible checkpoint_oldest_ts is 0 and checkpoint_ts is not 0. If we take the minimum first, we will not pin timestamp 0 but actually we should.

kommiharibabu replied:
I understood your point, but 0 is not a valid timestamp, so I feel the scenario of having one value as 0 and another not possible.

quchenhao replied:
It's possible in the first checkpoint that oldest is still 0. We may have stable timestamp at 20 and oldest still at 0. If we only read oldest, we can't decide whether we need to pin it or not. Also we may start a checkpoint without a timestamp. ```
    /* Clear a checkpoint's pinned ID. */
    if (WT_SESSION_IS_CHECKPOINT(session)) {
        txn_global->checkpoint_txn_shared.pinned_id = WT_TXN_NONE;
        txn_global->checkpoint_timestamp = 0;
    }
```
Here we only clears checkpoint_timestamp not oldest_ts. May be we should do that as well. Anyway, in neither case, we can solely rely on the oldest timestamp itself to decide if we need to pin or not.

kommiharibabu replied:
ok. Thanks for explaining.

quchenhao commented:
@agorrod @sueloverso @kommiharibabu @bvpvamsikrishna The csuite test failure for logged table is caused by WT-6637. I disabled the test for logged table for now. This change is now ready for review.

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,383 @@
+          session->create(session, uri, "key_format=S,value_format=u,log=(enabled=true)"));
+    else
+        testutil_check(
+          session->create(session, uri, "key_format=S,value_format=u,log=(enabled=false)"));
```
While I know that WT-6637 is open for the logged table, don't we really want both of these, all the time? It would seem that we want a logged table to simulate the oplog and a non-logged table to simulate a collection. Then the worker thread would work in both tables. I believe oplog and collection operations are part of the same transaction in MongoDB. If not, this should simulate whatever it is that MongoDB does.

agorrod replied:
I'm with @sueloverso here - it would be best if we can simulate MongoDB expectations, and match them here within the purview of this change.

We should document the expectations around checkpoints and recovery when logging tables and using timestamps, if that hasn't been done somewhere already. My belief is that recovery of logged tables with timestamps works as MongoDB expects - but we definitely need to make sure there aren't corner cases we don't understand.

quchenhao replied:
@agorrod  @sueloverso The problem doesn't impact mongodb as mongodb 4.4 only requires the logged table to have a consistent snapshot data at last successful commit after recovery. It doesn't read previous versions of data with timestamps. I think the issue here is that we haven't decided what is the expected behavior of logged table using timestamps after recovery. For non-logged table, we want to maintain a window of data from the oldest timestamp to the stable timestamp in the checkpoint. For logged table, do we need to maintain a window of data from the oldest timestamp to the last successful commit timestamp? If we don't, there is no point testing logged table in this test as it is to ensure we can read data of any version in the maintained time window after recovery. My view is that we need to maintain the window for logged table as well, otherwise the semantic for logged table will be very weird. Though we have a stable and oldest timestamp in the checkpoint, we can't actually read the data with timestamps for logged table. If we want to fix this, we need to record timestamps in the log record and replay them with timestamps. It is a data format change and will take quite some time to get right.

Meanwhile, this ticket is blocking us from merging the rts-test-format feature branch back to develop. In that feature branch, we have some changes that we desire to go to 4.4. Therefore, I prefer it to be merged soon. For now, we can either disable the log test or remove it for now until we know what is the expected hehavior for logged tables that use timestamps.

sueloverso replied:
Could we change the verification piece of this program, with a comment describing the mongodb usage and test the logged and non-logged data differently? That is my preference. Or if a logged table has different expectations, why have the `-l` arg in this test at all? Perhaps add a comment describing this and state the case for non-logged only.

sueloverso commented on test/csuite/wt6616_checkpoint_oldest_ts/main.c:
```diff
@@ -0,0 +1,383 @@
+            working_dir = __wt_optarg;
+            break;
+        case 'l':
+            log_table = true;
```
If you go with both an oplog-like and collection-like table, this arg would go away.

quchenhao commented:
@sueloverso @bvpvamsikrishna @agorrod There are two issues involved in WT-6637. The first issue is that log is not replayed from the correct point. The second issue is that all the data after the stable timestamp are lost because log are replayed without timestamps. It is enough to get the test green if we can fix the first one. However, fixing the second issue involves a data format change to record timestamp in logging, which we should be careful. I think I'll fix the first issue and merge this change.

quchenhao commented:
> @sueloverso @bvpvamsikrishna @agorrod There are two issues involved in [WT-6637](https://jira.mongodb.org/browse/WT-6637). The first issue is that log is not replayed from the correct point. The second issue is that all the data after the stable timestamp are lost because log are replayed without timestamps. It is enough to get the test green if we can fix the first one. However, fixing the second issue involves a data format change to record timestamp in logging, which we should be careful. I think I'll fix the first issue and merge this change.

I created WT-6639 for the first issue. However, after discussing with @kommiharibabu, I realized that it only alleviates the problem not fixing it (for details, please see the comment in WT-6639). To properly fix it, we still need to support timestamps in logging. Therefore, I'll leave the change as it is for now.

quchenhao commented on src/txn/txn.c:
```diff
@@ -131,10 +131,11 @@ __wt_txn_release_snapshot(WT_SESSION_IMPL *session)
         txn_global->checkpoint_txn_shared.pinned_id = WT_TXN_NONE;
-        txn_global->checkpoint_timestamp = 0;
+        txn_global->checkpoint_oldest_timestamp = WT_TS_NONE;
+        txn_global->checkpoint_timestamp = WT_TS_NONE;
```
I'm not sure do we need to ensure checkpoint_oldest_timestamp is cleared before checkpoint_timestamp using a write barrier?

quchenhao replied:
I think it is fine to not ensure the sequence here. If checkpoint_timestamp is cleared first and we have a race, we won't pin checkpoint_oldest_timestamp as we read checkpoint_timestamp as 0. If we clear checkpoint_oldest_timestamp first and we have a race, we read non-zero checkpoint_timestamp and pin zero timestamp (checkpoint_oldest_timestamp), which is conservative but still correct.

quchenhao commented:
@sueloverso I don't know what is the expected behavior for logged table when I wrote the test. Now I agree we should remove it from the test for now until we define it clearly.

bvpvamsikrishna reviewed (approved):
lgtm

kommiharibabu reviewed (approved):
lgtm

agorrod reviewed (commented):
I feel like this solution is at the wrong end of the problem. We don't (I think) guarantee that a checkpoint will keep the oldest timestamp as it was when it began. I think instead the checkpoint oldest timestamp should be a calculation that is `min(oldest, checkpoint)` timestamps as they are when checkpoint is being finished.

I stopped doing a detailed review once I understood the change.

quchenhao commented on src/include/txn.i:
```diff
@@ -510,15 +510,24 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
+     * checkpoint is starting and we have to use the oldest timestamp stored in the checkpoint, we
+     * take the minimum of it with the oldest timestamp, which is what we want.
+     */
+    WT_READ_BARRIER();
```
This read barrier is in the code before the change.
I think it is initially to guard against reordering `*pinned_tsp = pinned_ts = txn_global->pinned_timestamp;`

Later we added

```
if (!include_checkpoint_txn)
        return;
```
and now it is no longer required. I agree we can remove this.

quchenhao commented:
> I feel like this solution is at the wrong end of the problem. We don't (I think) guarantee that a checkpoint will keep the oldest timestamp as it was when it began. I think instead the checkpoint oldest timestamp should be a calculation that is `min(oldest, checkpoint)` timestamps as they are when checkpoint is being finished.
> 
> I stopped doing a detailed review once I understood the change.

That's a good idea. I'll try that.

agorrod commented on src/include/txn.i:
```diff
@@ -508,13 +508,6 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
-     * checkpoint is starting and we have to use the checkpoint timestamp, we take the minimum of it
-     * with the oldest timestamp, which is what we want.
-     */
-    WT_READ_BARRIER();
```
It looks to me like this read barrier is necessary. The comment says that we need to make sure that the read of the checkpoint_timestamp sees the value at least as recent as the read of the pinned_timestamp above. The compiler can re-order across a conditional, so I don't think that helps.

The code would be more obvious if the operations were grouped together. Could the check for `include_checkpoint_txn` be moved above assigning the pinned timestamp? So the two assignments where order matters are grouped together? I'd make it:

```
/*
 * The read of the timestamp pinned by a checkpoint needs to be carefully ordered: if a
 * checkpoint is starting and we have to use the checkpoint timestamp, we take the minimum of it
 * with the oldest timestamp, which is what we want.
 */
*pinned_tsp = pinned_ts = txn_global->pinned_timestamp;
WT_READ_BARRIER();
checkpoint_ts = txn_global->checkpoint_timestamp;
```

That makes the grouping more obvious, so hopefully we won't add code in the middle accidentally again.

quchenhao replied:
I think it should also be behind
```
include_checkpoint_txn =
      btree == NULL || (btree->checkpoint_gen != __wt_gen(session, WT_GEN_CHECKPOINT));
```
Otherwise, we may read the checkpoint timestamp from the previous checkpoint.

agorrod commented on src/meta/meta_ckpt.c:
```diff
@@ -918,9 +918,14 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
+     * and the checkpoint timestamp.
+     */
+    __wt_timestamp_to_hex_string(WT_MIN(S2C(session)->txn_global.oldest_timestamp,
+                                   S2C(session)->txn_global.meta_ckpt_timestamp),
```
This function now uses `S2C` in a number of places - it's better to make a local `conn` variable.

agorrod commented on src/meta/meta_ckpt.c:
```diff
@@ -918,9 +918,14 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
+     */
+    __wt_timestamp_to_hex_string(WT_MIN(S2C(session)->txn_global.oldest_timestamp,
+                                   S2C(session)->txn_global.meta_ckpt_timestamp),
+      hex_timestamp);
```
The indenting looks strange here - is it really the way it's supposed to be?

agorrod commented on src/meta/meta_ckpt.c:
```diff
@@ -918,9 +918,14 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
+     * should set the checkpoint's oldest timestamp as the minimum of the current oldest timestamp
+     * and the checkpoint timestamp.
+     */
+    __wt_timestamp_to_hex_string(WT_MIN(S2C(session)->txn_global.oldest_timestamp,
```
I'm a bit uncomfortable about this. I *think* the shared `txn_global::oldest_timestamp` could be read twice as part of this operation, and if that happened, we might store a value newer than `meta_ckpt_timestamp`, and I think that would be unexpected.
I believe you need a read of oldest into a local variable, followed by a read barrier, then the comparison to be safe here.

quchenhao replied:
I'm not sure what you mean. These operations are at the final stage of a checkpoint. Is it possible another checkpoint can start concurrently?

quchenhao replied:
Oh, I get it. Yes, you are correct. I don't know it may be read twice.

quchenhao replied:
I think I get it. But I think the read_barrier is not necessary.

quchenhao commented on src/meta/meta_ckpt.c:
```diff
@@ -918,9 +922,16 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
+     *
+     * Cache the oldest timestamp to prevent us from reading two different values.
+     */
+    oldest_timestamp = txn_global->oldest_timestamp;
```
I don't believe we need a read barrier here. The usage of oldest_timestamp happens after the definition so the compiler should not reorder that. `meta_ckpt_timestamp ` cannot change as we are still in a checkpoint.

agorrod replied:
The compiler will very likely optimize out the local variable in this case. That's what I suggested the barrier would prevent.

quchenhao replied:
OK, that makes sense.

quchenhao replied:
@agorrod Though I put a read barrier here, I actually prefer to declare the variable as volatile as we are not trying to guarantee the order here but merely to prevent it from being optimized away.

In addition, this reminds me the `txnid` local variable in `__rec_upd_select` may suffer the same problem.

agorrod replied:
We've oscillated on the `volatile` keyword, but moved away from using it, since the actual guarantees of volatile are not very useful. So, when we have a case like this we prefer to use barriers to enforce the expected behavior if possible.

agorrod commented on src/meta/meta_ckpt.c:
```diff
@@ -918,9 +922,18 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
+     * Cache the oldest timestamp to prevent us from reading two different values.
+     */
+    oldest_timestamp = txn_global->oldest_timestamp;
+    /* Add a read barrier to prevent the oldest_timestamp from being optimized away. */
```
It's not just being optimized away, it's that along with being read twice with different values.

agorrod reviewed (commented):
Please update the comment and run stress testing and MongoDB testing to make sure this change is OK. I believe there are cases where MongoDB was sensitive to getting the oldest timestamp old enough here.

quchenhao commented:
@agorrod I ran the mongodb test and stress test before and I'll run again with the latest change.

quchenhao commented:
@agorrod Mongodb tesets pass and there are two failures in the stress tests (WT-6504 and WT-6394)

kommiharibabu reviewed (approved):
lgtm

agorrod reviewed (approved):
lgtm
